rclcpp/include/rclcpp/buffers/intra_process_buffer.hpp

```diff
@@ -0,0 +1,208 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__BUFFERS__INTRA_PROCESS_BUFFER_HPP_
+#define RCLCPP__BUFFERS__INTRA_PROCESS_BUFFER_HPP_
+
+#include <memory>
+#include <type_traits>
+#include <utility>
+
+#include "rclcpp/buffers/buffer_implementation_base.hpp"
+
+namespace rclcpp
+{
+namespace intra_process_buffer
+{
+
+class IntraProcessBufferBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(IntraProcessBufferBase)
+
+  virtual void clear() = 0;
+
+  virtual bool has_data() const = 0;
+  virtual bool use_take_shared_method() const = 0;
+};
+
+template<typename MessageT>
+class IntraProcessBuffer : public IntraProcessBufferBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(IntraProcessBuffer)
+
+  using ConstMessageSharedPtr = std::shared_ptr<const MessageT>;
+  using MessageUniquePtr = std::unique_ptr<MessageT>;
```

Comment:
Why not `std::unique_ptr<const MessageT>` ?

Reply:
Because `std::unique_ptr<MessageT>` is the one that can be used in the callbacks
https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/any_subscription_callback.hpp#L40